File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findMinEleInRightSubtree(TreeNode node) {
        while(node.left!=null) node = node.left;
        return node.val;
    }
    public TreeNode deleteNode(TreeNode root, int key) {
        //Base Case
        if(root == null) return null;
        
        //binary search
        if(root.val < key) {
            //recurse to right side
            root.right = deleteNode(root.right,key);
        }
        else if(root.val > key) {
            root.left = deleteNode(root.left,key);
        }
        else{
            // Condition to handle both child null , right child null & left child null
            if(root.left == null || root.right == null) {
                return root.left == null ? root.right : root.left;
            }
            
            //if both children are present of node to be deleted
            //get min element in right subtree
            int minEle = findMinEleInRightSubtree(root.right);
            //as this element is min in our right subtree there is no chance that it will have 2 children or even 1 left child
            //substitute the val of min element in our current root.
            root.val = minEle;
            //delete the min element
            root.right = deleteNode(root.right,minEle);
        }
        return root;
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findMinInLeftSubtre(TreeNode* node) {
        while(node->left) node = node->left;
        //we get the minimum element in right subtree.
        return node->val;
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        //Base Case
        if(!root) return nullptr;
        //search in right
        if(root->val < key) {
            root->right = deleteNode(root->right,key);
        }
        //search in left
        else if(root->val > key) {
            root->left = deleteNode(root->left,key);
        }
        //else we have got the node to be deleted
        else {
            //if left child of node to be delete is null || right child is null or both are null
            if(root->left == nullptr || root->right == nullptr) {
                return root->left == nullptr ? root->right : root->left;
            }
            //both children left and right are present
            //get min element in right subtree
            int minEle = findMinInLeftSubtre(root->right);
            //subtitute our root's val with minEle
            root->val = minEle;
            //delete the node wil min element in right subtree
            root->right = deleteNode(root->right,minEle);
        }
        return root;
    }
};